static_frame/core/frame.py

```diff
@@ -1493,42 +1495,66 @@ def from_sql(cls,
-        elif columns_depth > 1:
-            # use IH: get via static attr of columns const
-            constructor = cls._COLUMNS_HIERARCHY_CONSTRUCTOR.from_labels_delimited
-            labels = (b[0] for b in row_gen.description[index_depth:])
-            columns = constructor(labels, delimiter=' ')
+        own_columns = False
+        filter_row = lambda row: row
+
+        if columns_depth >= 1:
+            labels = (col for (col, *_) in sql_iterable.description[index_depth:])
             own_columns = True
 
+            if columns_depth == 1:
+                if columns_select:
+                    indices, labels = map(tuple, zip(*((i, label) for (i, label) in enumerate(labels) if label in columns_select)))
+                    selector = itemgetter(*indices)
+                    filter_row = lambda row: selector(row)
+
+                columns = cls._COLUMNS_CONSTRUCTOR(labels)
+            else: # > 1
+                # use IH: get via static attr of columns const
+                constructor = cls._COLUMNS_HIERARCHY_CONSTRUCTOR.from_labels_delimited
+                columns = constructor(labels, delimiter=' ')
+
+                if columns_select:
+                    # Have to wait until columns is already constructed in order to filter, since
+                    # `from_labels_delimited` is implementing specific logic to convert the SQL columns
+                    # into an IndexHierarchy, thus making filtering on the raw labels effectively impossible
+                    columns_select = set(columns_select)
+                    sel = columns.isin(columns_select)
+                    columns = columns.loc[sel]
+
+                    selector = itemgetter(*np.where(sel)[0])
+                    filter_row = lambda row: selector(row)
+
         index_constructor = None
 
         if index_depth > 0:
             index = [] # lazily populate
             if index_depth == 1:
                 index_constructor = Index
 
-                def row_gen_final() -> tp.Iterator[tp.Sequence[tp.Any]]:
-                    for row in row_gen:
+                def row_gen() -> tp.Iterator[tp.Sequence[tp.Any]]:
+                    for row in sql_iterable:
                         index.append(row[0])
                         yield row[1:]
 
             else: # > 1
                 index_constructor = IndexHierarchy.from_labels
 
-                def row_gen_final() -> tp.Iterator[tp.Sequence[tp.Any]]:
-                    for row in row_gen:
+                def row_gen() -> tp.Iterator[tp.Sequence[tp.Any]]:
+                    for row in sql_iterable:
                         index.append(row[:index_depth])
                         yield row[index_depth:]
         else:
             index = None
-            row_gen_final = lambda: row_gen
+            row_gen = lambda: sql_iterable
+
+        if columns_select:
+            row_gen_final = lambda: (filter_row(row) for row in row_gen())
+        else:
+            row_gen_final = row_gen
```

Comment:
Rather then pack the iterators behind `lambda`, maybe just call `row_gen` here (line 1557) and then do not call it in line 1561. That will save some unnecessary lambda calls, as well as assigning the lambda to a name.